management/src/main/java/org/ehcache/management/cluster/DefaultClusteringManagementService.java

```diff
@@ -0,0 +1,178 @@
+
+import java.util.Collection;
+
+@ServiceDependencies({CacheManagerProviderService.class, ExecutionService.class, TimeSourceService.class, ManagementRegistryService.class, EntityService.class})
+public class DefaultClusteringManagementService implements ClusteringManagementService, CacheManagerListener, CollectorService.Collector {
+
+  private volatile ManagementRegistryService managementRegistryService;
+  private volatile CollectorService collectorService;
+  private volatile ManagementAgentService managementAgentService;
+  private volatile ClientEntityFactory<ManagementAgentEntity, ManagementAgentConfig> managementAgentEntityFactory;
+  private volatile InternalCacheManager cacheManager;
+
+  @Override
+  public void start(ServiceProvider<Service> serviceProvider) {
+    this.managementRegistryService = serviceProvider.getService(ManagementRegistryService.class);
+    this.cacheManager = serviceProvider.getService(CacheManagerProviderService.class).getCacheManager();
+
+    this.collectorService = new DefaultCollectorService(this);
+    this.collectorService.start(serviceProvider);
+
+    EntityService entityService = serviceProvider.getService(EntityService.class);
+    this.managementAgentEntityFactory = entityService.newClientEntityFactory(
+        "ManagementAgent",
+        ManagementAgentEntity.class,
+        Version.LATEST.version(),
+        new ManagementAgentConfig()
+            //TODO: this config is only used once when the management entity is created. It's useless otherwise.
+            // should-we add something in ehcache config to handle the creation case ?
+            .setMaximumUnreadClientNotifications(1024 * 1024)
+            .setMaximumUnreadClientStatistics(1024 * 1024));
+
+    this.cacheManager.registerListener(this);
+  }
+
+  @Override
+  public void stop() {
+    collectorService.stop();
+
+    // nullify so that no further actions are done with them (see null-checks below)
+    managementRegistryService = null;
+    managementAgentService = null;
+  }
+
+  @Override
+  public void cacheAdded(String alias, Cache<?, ?> cache) {
+  }
+
+  @Override
+  public void cacheRemoved(String alias, Cache<?, ?> cache) {
+  }
+
+  @Override
+  public void stateTransition(Status from, Status to) {
+    // we are only interested when cache manager is initializing (but at the end of the initialization)
+    switch (to) {
+
+      case AVAILABLE: {
+        // create / fetch the management entity
+        ManagementAgentEntity managementAgentEntity;
+        try {
+          managementAgentEntity = managementAgentEntityFactory.retrieve();
+        } catch (EntityNotFoundException e) {
+          try {
+            managementAgentEntityFactory.create();
+          } catch (EntityAlreadyExistsException ignored) {
+          }
+          try {
+            managementAgentEntity = managementAgentEntityFactory.retrieve();
+          } catch (EntityNotFoundException bigFailure) {
+            throw new AssertionError(bigFailure);
```

Comment:
`AssertionError` means bug, hence I'd include a human-readable error message so that one would have a rough idea of why the exception was thrown in case we ever run into that bug.